### PR TITLE
API for granting support access and registering granted certificates

### PR DESCRIPTION
**Pull request by andreer**
…cates

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccessControl.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            if (until.isAfter(now.plus(MAX_SUPPORT_ACCESS_TIME))) {
+                throw new IllegalArgumentException("Support access cannot be allowed for more than 10 days");
+            }
+            SupportAccess allowed = forDeployment(deployment).withAllowedUntil(until, by, now);
+            controller.curator().writeSupportAccess(deployment, allowed);
+            return allowed;
+        }
+    }
+
+    public SupportAccess registerGrant(DeploymentId deployment, String by, X509Certificate certificate) {
```
Comment (andreer) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccessControl.java:60]: Unused - will be used by @tokle

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccess.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
+package com.yahoo.vespa.hosted.controller.support_access;
```
[the remark this answers is not in the document]
Comment (oyving) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccess.java:1]: Also, copyright notice

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccess.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class SupportAccess {
```
Comment (oyving) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccess.java:10]: Javadoc. Should say if it's immutable or not.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccess.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    private final List<SupportAccessChange> changeHistory;
+    private final List<SupportAccessGrant> grantHistory;
+
+    /**
+     * public for serializer - do not use
+     */
+    public SupportAccess(List<SupportAccessChange> changeHistory, List<SupportAccessGrant> grantHistory) {
+        this.changeHistory = changeHistory;
+        this.grantHistory = grantHistory;
```
Comment (oyving) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccess.java:22]: If `SupportAccess` is immutable, these lists should be immutable as well. Either here or in the getters exposing them.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccessChange.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
+package com.yahoo.vespa.hosted.controller.support_access;
+
+import java.time.Instant;
+import java.util.Objects;
+import java.util.Optional;
+
+public class SupportAccessChange {
```
Comment (oyving) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccessChange.java:7]: javadoc and copyright

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccessControl.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
+package com.yahoo.vespa.hosted.controller.support_access;
```
Comment (oyving) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/support_access/SupportAccessControl.java:1]: copyright

**Review by oyving — CHANGES_REQUESTED** [review]
Some copyright notices are missing on the new files. A package should probably be renamed. Some missing javadoc.

**Review by oyving — APPROVED** [review]

**Pull request merged**